[QNNPACK] Block Sparse kernel. First commit.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51120 [QNNPACK Sparsity] Add aarch64 kernel of 8x1 sparsity
* #51119 [QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32.
* #51118 [QNNPACK, Sparsity] Code refactoring to allow for more generic block
* #50591 [QNNPACK, Sparsity] Add dyanmic linear sparse kernel for arm64
* #50590 [QNNPACK, Sparsity] Sparse kernel with 4x8 blocking
* #50589 [QNNPACK, Sparsity] Added prepacking base aarch32 kernels
* #50588 [QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear
* #50587 [QNNPACK:Sparsity] Add A matrix pretransformed based sparse kernels for FC
* #50586 [QNNPACK Sparse] Create fc sparse operator
* **#50585 [QNNPACK] Block Sparse kernel. First commit.**

Summary:
This diff introduces sparse kenel for sse. Uses 1x4 block sparse
pattern.

Test Plan:
./build/local/q8gemm-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925500](https://our.internmc.facebook.com/intern/diff/D25925500)

kimishpatel committed:
[QNNPACK] Block Sparse kernel. First commit.

Summary:
This diff introduces sparse kenel for sse. Uses 1x4 block sparse
pattern.

Test Plan:
./build/local/q8gemm-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/pack_block_sparse.h:
```diff
@@ -0,0 +1,106 @@
+    const uint8_t* zero_points) {
+  BCSRMatrix bcsr_mat;
+  bcsr_mat.row_values.resize(N);
+  // K must be > 0
```
An assert would be helpful to check for preconditions.  I think generous use of asserts is good practice - has no runtime cost but can come in handy when dealing with code that behaves in unexpected ways.

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/pack_block_sparse.h:
```diff
@@ -0,0 +1,106 @@
+    (K % col_block_size) == 0 ? num_blocks : num_blocks - 1;
+  const uint32_t remainder_elements = K % col_block_size;
+
+  bcsr_mat.row_values.clear();
```
bcsr_mat.row_values.resize() followed by clear should not be necessary when followed by reserve() and operating on a newly constructed vector.  If you just want to pre-allocate memory without modifying the vector size then a call to reserve should be enough.

kimishpatel replied:
Yeah this is an oversight. Thanks for catching. Will fix it.

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/test/gemm-block-sparse-microkernel-tester.h:
```diff
@@ -0,0 +1,346 @@
+#include <qnnpack/params.h>
+#include <qnnpack/requantization.h>
+
+namespace {
```
This doesn't matter since this is a unit test but placing anonymous or static functions in a header file means each compilation unit is going to end up with its own copy since this implies internal linkage.  I am not actually sure if the compiler consolidates all [similar] functions into one in practice but avoid this usage personally.

kimishpatel replied:
I think you are right. It will just create a separate copy for each translation unit. But I am more lax with test code and did this just out of practice.

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm_sparse/8x4c1x4-dq-sse2.c:
```diff
@@ -0,0 +1,442 @@
+
+#define MR 8
+#define COL_BLOCK_SIZE 4
+#define PACKED_A_BLOCK_SIZE COL_BLOCK_SIZE*MR
```
This doesn't to be used anywhere.

kimishpatel replied:
Thanks for the catch.

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm_sparse/8x4c1x4-dq-sse2.c:
```diff
@@ -0,0 +1,442 @@
+#define COL_BLOCK_SIZE 4
+#define PACKED_A_BLOCK_SIZE COL_BLOCK_SIZE*MR
+
+#define LOAD_TRANSPOSED_A_BLOCK(id, index, col_block_id)                  \
```
I personally lean on using forced inline functions if possible.  The thing with macros like this is that if you end up accidentally using any arithmetic in the arguments at call-site LOAD_TRANSPOSED_A_BLOCK(a, b, c+1) this will have unexpected consequences.  Also the commonality like `col_block_id * COL_BLOCK_SIZE + index` can be factored out more readily when using functions.

kimishpatel replied:
I opted this way mainly because I can generate variables names like va0, va1 etc. that are later used and can avoid passing 10 args to the function. I agree about arithmetics used in arg leading to unexpected consequences. I will fix it appropriately.

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/CMakeLists.txt:
```diff
@@ -571,6 +572,15 @@ if(PYTORCH_QNNPACK_BUILD_TESTS)
 
+  add_executable(q8gemm-sparse-test test/q8gemm_sparse.cc)
+  set_target_properties(q8gemm-sparse-test PROPERTIES
+    CXX_STANDARD 14
```
I take this back, I was looking at the original file. :)

kimishpatel committed:
Update on "[QNNPACK] Block Sparse kernel. First commit."


Summary:
This diff introduces sparse kenel for sse. Uses 1x4 block sparse
pattern.

Test Plan:
./build/local/q8gemm-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925500](https://our.internmc.facebook.com/intern/diff/D25925500)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK] Block Sparse kernel. First commit."


Summary:
This diff introduces sparse kenel for sse. Uses 1x4 block sparse
pattern.

Test Plan:
./build/local/q8gemm-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925500](https://our.internmc.facebook.com/intern/diff/D25925500)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK] Block Sparse kernel. First commit."


Summary:
This diff introduces sparse kenel for sse. Uses 1x4 block sparse
pattern.

Test Plan:
./build/local/q8gemm-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925500](https://our.internmc.facebook.com/intern/diff/D25925500)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK] Block Sparse kernel. First commit."


Summary:
This diff introduces sparse kenel for sse. Uses 1x4 block sparse
pattern.

Test Plan:
./build/local/q8gemm-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925500](https://our.internmc.facebook.com/intern/diff/D25925500)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK] Block Sparse kernel. First commit."


Summary:
This diff introduces sparse kenel for sse. Uses 1x4 block sparse
pattern.

Test Plan:
./build/local/q8gemm-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925500](https://our.internmc.facebook.com/intern/diff/D25925500)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK] Block Sparse kernel. First commit."


Summary:
This diff introduces sparse kenel for sse. Uses 1x4 block sparse
pattern.

Test Plan:
./build/local/q8gemm-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925500](https://our.internmc.facebook.com/intern/diff/D25925500)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK] Block Sparse kernel. First commit."


Summary:
This diff introduces sparse kenel for sse. Uses 1x4 block sparse
pattern.

Test Plan:
./build/local/q8gemm-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925500](https://our.internmc.facebook.com/intern/diff/D25925500)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK] Block Sparse kernel. First commit."


Summary:
This diff introduces sparse kenel for sse. Uses 1x4 block sparse
pattern.

Test Plan:
./build/local/q8gemm-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925500](https://our.internmc.facebook.com/intern/diff/D25925500)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK] Block Sparse kernel. First commit."


Summary:
This diff introduces sparse kenel for sse. Uses 1x4 block sparse
pattern.

Test Plan:
./build/local/q8gemm-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925500](https://our.internmc.facebook.com/intern/diff/D25925500)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK] Block Sparse kernel. First commit."


Summary:
This diff introduces sparse kenel for sse. Uses 1x4 block sparse
pattern.

Test Plan:
./build/local/q8gemm-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925500](https://our.internmc.facebook.com/intern/diff/D25925500)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK] Block Sparse kernel. First commit."


Summary:
This diff introduces sparse kenel for sse. Uses 1x4 block sparse
pattern.

Test Plan:
./build/local/q8gemm-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925500](https://our.internmc.facebook.com/intern/diff/D25925500)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK] Block Sparse kernel. First commit."


Summary:
This diff introduces sparse kenel for sse. Uses 1x4 block sparse
pattern.

Test Plan:
./build/local/q8gemm-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925500](https://our.internmc.facebook.com/intern/diff/D25925500)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK] Block Sparse kernel. First commit."


Summary:
This diff introduces sparse kenel for sse. Uses 1x4 block sparse
pattern.

Test Plan:
./build/local/q8gemm-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925500](https://our.internmc.facebook.com/intern/diff/D25925500)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK] Block Sparse kernel. First commit."


Summary:
This diff introduces sparse kenel for sse. Uses 1x4 block sparse
pattern.

Test Plan:
./build/local/q8gemm-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925500](https://our.internmc.facebook.com/intern/diff/D25925500)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK] Block Sparse kernel. First commit."


Summary:
This diff introduces sparse kenel for sse. Uses 1x4 block sparse
pattern.

Test Plan:
./build/local/q8gemm-sparse-test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925500](https://our.internmc.facebook.com/intern/diff/D25925500)

[ghstack-poisoned]
